### DISCUSSION TITLE
Adding CT's

### DISCUSSION THREAD

**Discussion by n4ziq** [jgyates/genmon#1055] [Questions] [answered]
Has anyone added CT's (Current Transducers/Transformer) to a Generac Water Cooled unit? My 27Kw has no option on the wiring diagram for a CT. I just can't add one and input it into the existing Evo controller. I haven't seen any recent Pi HAT's that interface with Genmon that will work, unless I missed something. I see some posts/things from 2020, but the hardware referenced in 2020 doesn't seem to be available anymore. Thanks, Joe

**Comment by jgyates** [comment]
I don't think pintsize.me makes the HAT CT any more, but @skipfire can weigh in on that.

I use external CT, but in a slightly different way than the HAT CT board. At the moment I am the only one using it (that I know of) this way as I don't have enough info to make it an add on or more open interface. Here is what I am doing:

I have an open energy monitor EmonPi that monitors my home energy usage:

     https://docs.openenergymonitor.org/emonpi/technical.html

I have two CTs that are connected to the load side of my transfer switch. This allows it to report on energy usage regardless if there is an outage or not. I have a python program that calls into the emonpi's interface. Genmon will use the Hall Effect Sensor to report current if the transfer switch is not active (exercise cycles), if the transfer switch is active it will use the value imported from the EmonPi API. 

I am open to making a plug in or publising an API that allows 3rd party CTs. The current (undocumented) genmon interface allows either power or current to be imported.

Another complicating factor, Open Energy Monitor has changed their design a bit since I purchased the Emonpi. Now they offer what looks like an improved way to monitor:  https://shop.openenergymonitor.com/energy-monitoring/. When I setup mine I had to replace a resistor to get a larger rated CT, but it looks like the new version have more flexibility. I don't know if their API has changed since I wrote my code. 

Let me know if you have any questions.

**Reply by skipfire** [reply]
I did discontinue them due to problems with both setup and accuracy.

I have IoTaWatt (based off OEM) at my house and also have clamps on my generator feed as well as many other circuits (I actually have 4 IoTaWatts). I'd love to see an option for integrating that data and I'd be happy to be a second reference point if needed @jgyates.

(previously left comment from the wrong GitHub account, replaced it on the correct one)

**Comment by jgyates** [comment]
is there a public api for accessing iotawatt data? Is is stored local on the device or is it in the cloud?

**Reply by jgyates** [reply]
looks like Iotawatt is not producing boards any longer: https://community.iotawatt.com/t/not-ending-retail-sales/5654

**Reply by skipfire** [reply]
There is an API available on-device. I can get some sample request & data put together tonight or tomorrow morning once I'm home. As I understand it is built upon OEM but hardware made for the US so the calls may be the same. We can also sync through email and I can probably get you remote access to one of mine so you have data to work against.

(previously left comment from the wrong GitHub account, replaced it on the correct one)

**Reply by skipfire** [reply]
That post about IoTaWatt discontinuing sales is interesting since it also says they are not ending retail sales in that post.  Since it was based on OpenEnergyMonitor and OpenEnergyMonitor has US kits now, I'd say use that rather than IoTaWatt.

**Reply by jgyates** [reply]
I am thinking about have a couple of options, one like OpenEnergyMonitor, and another that will subscribe to a MQTT broker with power or current data. Any ideas / suggestions are welcome.

@n4ziq, what version is your firmware in your Liquid cooled model. I understand that there can be some some variance between the modbus output for current between firmware variants. If you have outbound email setup send me your logs if you can. I have a CT and calibrated my genmon readings today. One the advanced settings page I used a Current Divider of 33 and and Current Offset of 0. This makes my Hall Effect Sensor output (the default way genmon reads current over modbus for liquid cooled Evlution units) right in line with my CTs. Your milage may vary, but it is worth a shot.

**Reply by n4ziq** [reply]
I will try the current divider at 33 with offset of 0. 
I had a current divider of 0.5 and offset of -334. Why I have those values, I can't remember. It's been a while since I put an ammeter on the load side and calculated the scaling values. 

Hardware version is 1.04
Firmware version is 1.36

**Comment by jgyates** [comment]
I got your logs, thanks. You may want to get a data point for having no divisor or offset. I added some logic that factors in your firmware version on handling this register so that may make a difference also. I belive this sensor (hall effect sensor, which detects magnetism), like the CTs on the air cooled units, must be calibrated during install of the generator,  so the results could vary wildly.

**Reply by n4ziq** [reply]
Thanks. Sorry for the delayed response, busy week. What is the best way to monitor register 0058 while measuring current? I brought up the register view and was not sure of the polling rate. I was hoping to be able to align the instantaneous current with the 0058 register to do some calculations. I noticed the graph when I clicked on the register, and it seemed the graph update more frequently than the Hex value, but I wasn't 100% sure.

**Comment by jgyates** [comment]
The polling rate of genmon is "as fast as it can" (e.g. a thread in a loop reading registers over and over), the polling rate of the web app is 5 seconds. The best way to view the registers is the live view. The best way to to try to get a stable current source like a hair dryer an oven on broil, then the few second lag in polling / display is minimized.

**Reply by n4ziq** [reply]
Just wanted to provide a quick update. I borrowed a second Ammeter so I was able to monitor both L1 and L2 simultaneously. It seems as register 0058 decreases from 180, the current goes up. 178 was 16 amps, 176 was 23 amps, 174 was 48 amps. Current divider of 0.125 and offset of -1440. Seems about right now... Need to do more validation, but having 2 ammeter made a big difference since I clearly do not have a balanced load on L1 and L2 in the house. When I first did this years ago, I used 1 ammeter and just doubled the number...

**Comment by Mehrdadg** [comment]
I use IOTAWATT (https://iotawatt.com/) along with MQTT to publish all the data to nodered. In nodered, I process the data and send notifications based on some triggers I wanted to notify me. Additionally, from nodered, every data point is also sent to an influxdB.

Using Grafana (great opensource visualization sw), I display all sorts of data coming from my IoTAwatt device.

Also IotaWatt does record the data into its SD card. Also worth mentioning the IOTA device is very reliable as it has been running for 3+ years without a single issue

By the way the owner of the IoTAwatt project, boblemaire, is as responsive and helpful as the owner of this project so if you have any questions and feature adds, he will respond right away.

**Comment by jgyates** [comment, marked as answer]
I have created a new add on program that will allow data to be imported into genmon via MQTT:

https://github.com/jgyates/genmon/wiki/1----Software-Overview#genmqttinpy-optional

Let me know if the wiki does not answer your questions.
